Feature: Per-profile config settings

markstos opened the pull request:
Adds per-profile config settings.

Fixes #11 using the design proposed there.

Example:

    qutebrowser-profile --new test --set "downloads.location.directory" /tmp

After running this, inspect the generated file in `~/.local/share/applications` to confirm
the `--set` args are being passed through `qutebrowser`.

Then launch the browser using the `.desktop` file:

    gtk-launch qutebrowser-test.desktop

Use `:set downloads.location.directory` to inspect the current value and confirm it
is `/tmp`.

Open `~/.config/qutebrowser/autconfig.yml` and confirm that downloads.location.directory has
not been overrriden in the main config, but is a session-specific value.

Commit:
Fix incorrect choose flag error, load variation.

Whatever broke that required PR #2 to fix calls to --new also affected
calls to --load. Here's the behavior before this patch:

Before:

> ./qutebrowser-profile --load "test"
> cannot use --choose with --load

After applying it, a new profile is loaded and launched as expected.

Commit:
feature: Add support for unique WM Ids and desktop files.

Fixes #7 and #8.

I don't see much downside to this, but note it depends on the version of
Qutebrowser which I think is not yet released.

As before, there is no "delete profile" feature, and adding .desktop
files does provide one thing to manually clean up.

This work depends on #9, so it appears in this PR as well.

Commit:
feature: per-profile configuration options

Now when you combine --set OPTION VALUE with --new, the --set calls will
be persisted in the generated `.desktop` file, allowing per-profile
config options.

markstos commented:
Because this builds on my other recent PRs, commits for them appear here as well.

pjvds commented:
Highly appreciated and well worth pulling in 👍🏻

I've been using the window title to distinguish between profiles and give them a unique border color.

markstos commented:
I found a quoting bug with this PR that needs to be worked out. It's triggered when a `--set` option value contains a space, like setting a custom title.

Commit:
WIP: fix quoting.

Not working yet. A test case looks like this:

    qutebrower-profile --load work --set window.title_format "{current_title} WORK"

When quoting is fixed, this should result in the args getting passeds
through the qutebrowse the same way:

    qutebrower --set window.title_format "{current_title} WORK"

markstos commented:
@bitmeise @pjvds I can't quite figure out how to fix the bash quoting/escaping problem. Could take a look at and see if you can fix the issue? I think it's a one line fix but I can't seem to figure it out.

991645aafbb5080507e0b3897c3953cee4f314cd

 A test case looks like this:

    qutebrower-profile --load work --set window.title_format "{current_title} WORK"

When quoting is fixed, this should result in the args getting passeds
through the qutebrowse the same way:

    qutebrower --set window.title_format "{current_title} WORK"

I've tried using Bash arrays, the "@Q" notation for quoting variables and `printf`, but I can't seem to work out the exact right combination. 

Thanks!

markstos commented:
It's also tempting to rewrite the whole thing in anything-but-bash just because this is an annoying bashism. 😄

markstos commented:
A rewrite in Python might make profile management that much closer to becoming an official `qutebrowser` feature...

bitmeise commented:
> @bitmeise @pjvds I can't quite figure out how to fix the bash quoting/escaping problem. Could take a look at and see if you can fix the issue? I think it's a one line fix but I can't seem to figure it out.

To be honest, I'm currently not a qutebrowser-profile user. I would support the reasoning though, that this is a welcomed feature for upstream qutebrowser.

Anyway, I tried to have a look, but have problems getting the thing to run at all. This is probably due to me having an unusual setup.
Sorry.

Commit:
Share userscripts / greasemonkey scripts between profiles.

"Userscripts" are often activated by keybinding configurations, so if
you are going to share your configs between profiles, it makes sense
to to share these as well.

markstos commented:
I've added a commit to share the userscripts and greasemonkey scripts between the profiles as well. 

If you tend to use different websites in different profiles, this will hurt nothing, and if you do use the same website in different profiles, you will likely appreciate havintg the same scripts to tweak them available across profiles.

Commit:
share dictionaries between profiles

markstos commented:
This branch has been updated to share dictionaries between profiles.

jtyers commented:
I wonder if a better approach for this would be to source `config.py` in the profile directory if it exists, rather than having users needing to set per-profile options on the command line. This removes issues around quoting when working with strings and means that your profile directory includes your entire qutebrowser config rather than it being partially embedded in a `.desktop` file.

The right answer is probably that both approaches have valid use cases!

markstos commented:
I only use `autoconfig` and not `config.py`. I suppose I could start using it, but it would prefer not to mix the two systems.

jtyers commented:
Thanks for your work on this @markstos. I've re-implemented `qutebrowser-profile` in Python which solves a few headaches with its implementation. Would it be possible to re-work your implementation in Python?

markstos commented:
My branch is working well-enough for me and I don't plan to re-implement this.
